### Business Contex

## Project Goals
Analyze testing datasets to identify correlations and opportunities for system improvements.
Provide data-driven recommendations to optimize operations and reduce costs.
Enhance participants’ skills in statistical analysis, data modeling, and communicating results to non-technical audiences.
Suggest outcomes from our DOE tests to increase system consistency

## Project Outcomes
Clear identification of process inefficiencies and performance bottlenecks.
Actionable recommendations to improve operational efficiency and productivity.
Stakeholder-ready reports and visualizations that drive informed decision-making.

## Project Deliverables

Cleaned and analyzed operational dataset with key performance metrics.
Statistical analysis report highlighting inefficiencies and root causes.
Visualizations and dashboards illustrating findings.
Final presentation with prioritized recommendations for process improvements.

## 8-Week Timeline (50 Hours per Participant)

- Week 1 – Project Kickoff & Data Understanding (5 hrs)
Define project scope, KPIs, and success criteria.
Review datasets, data dictionaries, and context for testing.
- Week 2 – Data Cleaning & Integration (6 hrs)
Clean and preprocess testing data.
Integrate multiple data sources as needed (e.g., testing logs).
- Week 3 – Exploratory Data Analysis (6 hrs)
Conduct statistical analysis to identify trends, anomalies, and bottlenecks.
Share initial findings for feedback.
- Week 4 – Metric Development & Hypothesis Testing (6 hrs)
Define key testing metrics (e.g., quality, consistency).
Test hypotheses to validate root causes of inefficiencies.
- Week 5 – Visualization & Reporting (7 hrs)
Develop visualizations (dashboards, charts) to represent findings.
Draft interim report with preliminary recommendations.
- Week 6 – Recommendation Development (6 hrs)
Refine insights and translate analysis into actionable recommendations.
Prioritize recommendations based on impact and feasibility.
- Week 7 – Final Report & Documentation (7 hrs)
Compile final report including methodology, analysis, and recommendations.
Prepare supporting documentation for long-term use.
- Week 8 – Presentation & Handover (7 hrs)
Present findings and recommendations to stakeholders.
Deliver final report, visualizations, and documentation package.

## **Planned Data Dictionary:**

The data dictionary is **the planned list of features**, so it is just listing what the **feature name** is, **what it means**, the
**theoretical equation for calculating it**, and what **type of metric it is**.

| Category | Planned Feature | Raw Data / Indicators Required | Mathematical Approach / Physiological Concept |
| :--- | :--- | :--- | :--- |
| **Pulmonary** | Alveolar Ventilation ($V_A$) | Tidal Volume ($V_T$), Dead Space ($V_D$), Respiratory Rate (RR) | Calculated volume of fresh air reaching the alveoli per minute: $$V_A = (V_T - V_D) \times RR$$ |
| **Cardiac** | Baroreflex Sensitivity (BRS) | HRV, Beat-to-beat blood pressure estimates | Evaluation of the coupling and directional latency between heart rate variability and continuous pressure fluctuations. |
| **Pulmonary** | Breathing Efficiency | Tidal Volume ($V_T$), Respiratory Rate (RR), Humidity | Algorithmic estimation of gas exchange optimization per breath cycle, factoring in environmental and exhaled moisture gradients. |
| **Pulmonary** | Breathing Frequency and Regularity | Breath timing data | Continuous monitoring of peak-to-peak interval variance to detect clinical irregularities (e.g., Cheyne-Stokes, apneas). |
| **Cardiac** | Cardiac Output (CO) | Heart Rate (HR), Stroke Volume (SV) | Continuous multiplication of heart rate and estimated stroke volume: $$CO = HR \times SV$$ |
| **Other** | Center of Mass & Stability | Acceleration, Tilt data | Biomechanical multi-axis vector analysis to determine positional equilibrium. |
| **Other** | Core Body Temperature (estimate) | Ambient temperature, Exhaled air temperature, Humidity | Thermodynamic heat transfer modeling to flag hyperthermia, hypothermia, and ambient heat stress risk. |
| **Pulmonary** | Dead Space Ventilation ($V_D$) | Tidal Volume ($V_T$), $CO_2$ kinetics | Extrapolating the volume of air not participating in gas exchange based on capnography clearance curves. |
| **Other** | Fluid Balance | *Not yet defined* | — |
| **Cardiac** | Heart Rate (HR) | PPG waveform | Time-domain tracking of systolic peaks from optical photoplethysmography arrays. |
| **Cardiac** | Heart Rate Variability (HRV) | PPG pulse intervals | Statistical evaluation of R-R (or peak-to-peak) interval variations to map autonomic nervous system (ANS) stress profiles. |
| **Other** | Hydration & Metabolic Stress Correlation | HR, Exhaled humidity, $CO_2$ | Multivariate stress indexing linking cardiovascular exertion, evaporative respiratory loss, and metabolic output. |
| **Other** | Hydration Status V | Exhaled/ambient moisture differences, Respiratory patterns, HR | Dynamic monitoring of physiological water loss via expired breath to trace dehydration trends. |
| **Pulmonary** | Inhaled Air Composition | Ambient humidity, Specific humidity | Sensor calibration Baseline checking of current atmospheric water vapor fractions. |
| **Pulmonary** | Inspiratory & Expiratory Flow Rate | Airflow data, Time | Calculation of the rate of volume displacement over time ($dV/dt$) to screen for obstructive or restrictive patterns. |
| **Pulmonary** | Lung Compliance & Elastance | Pressure, Volume | Delta volume divided by delta pressure ($\Delta V / \Delta P$) analyzed via systemic hysteresis loops. |
| **Other** | Metabolic Rate | Respiratory Exchange Ratio (RER) | Estimated energetic breakdown derived from gas consumption rates. |
| **Pulmonary** | Oxygen Delivery ($DO_2$) | Cardiac Output (CO), Arterial oxygen content ($SpO_2$, $O_2$) | Product of systemic fluid delivery and total carrier saturation values. |
| **Other** | Physical Activity Level (PAL) | 3-axis Acceleration | Classifying daily activity windows into functional tiers (sedentary, light, moderate, vigorous) using movement area scoring. |
| **Other** | Posture and Gait Analysis | Tilt, Acceleration | Angular displacement tracking over time to assess gait phase symmetry, structural balance, and musculoskeletal anomalies. |
| **Pulmonary** | Pulmonary Vascular Resistance (PVR) | $CO_2$, HR, $SpO_2$ | Indirect hemodynamic inference of right-ventricular afterload strains using downstream oxygenation and heart rate variables. |
| **Pulmonary** | Respiratory Exchange Ratio (RER) | $CO_2$ production ($VCO_2$), $O_2$ consumption ($VO_2$) | Metabolic metabolic calculation to determine active fuel usage substrate: $$RER = \frac{VCO_2}{VO_2}$$ |
| **Pulmonary** | Respiratory Muscle Strength/Endurance | Inspiratory/Expiratory pressures | Waveform thresholding of absolute peak pressures to quantify physiological fatigue or fatigue limits. |
| **Pulmonary** | Respiratory Rate (RR) | Volume or Pressure waves | Frequency count of breath cycles captured per discrete 60-second window. |
| **Other** | Respiratory Water Loss | Absolute ambient humidity, Absolute exhaled humidity | Mass balance calculation tracking net fluid mass lost per breath via absolute humidity differential profiles. |
| **Other** | Resting Metabolic Rate (RMR) | $VO_2$ at rest, Age, Gender, Weight | Combining steady-state resting oxygen consumption metrics with standard anthropometric predictive equations. |
| **Cardiac** | Stroke Volume (SV) | HR, $SpO_2$, $CO_2$ | Empirical mathematical extraction mapping variations in downstream volumetric trends from oxygenation and pulse contours. |
| **Other** | Sympathovagal Balance | HRV, Respiratory Rate, $CO_2$ levels | Multi-parametric scoring of the dynamic equilibrium existing between sympathetic (fight or flight) and parasympathetic systems. |
| **Other** | Thermoregulatory Response | *Not yet defined* | — |
| **Pulmonary** | $VCO_2$ (CO2 Output) | Exhaled $CO_2$, Tidal Volume ($V_T$), Respiratory Rate (RR) | Integration of the instantaneous product of exhaled flow volume and gas fraction configurations across breath boundaries. |
| **Pulmonary** | Ventilation-Perfusion (V/Q) Ratio | Capnography, Oxygenation trends | Mapping anomalies in blood perfusion relative to gaseous ventilation to check for structural ventilation mismatches. |
| **Pulmonary** | Ventilatory Efficiency | Gas exchange metrics | Slope index tracking gas utilization vs minute ventilation output. |
| **Pulmonary** | Ventilatory Equivalents ($VE/VO_2$) | Minute ventilation ($VE$), Estimated oxygen consumption ($VO_2$) | Mathematical verification of the volume of air needed to metabolize a fixed amount of oxygen. |
| **Pulmonary** | $VO_2$ (O2 Uptake) | Oxygen levels, Flow Rate, Time | Breath-by-breath integration mapping fractional oxygen drop against absolute transit volume. |
| **Pulmonary** | $VO_2\text{max}$ | $VO_2$, HR data, $VCO_2$ (RER) | Multi-point extrapolative estimation targeting maximum oxygen kinetics during metabolic peaks. |
| **Pulmonary** | Work of Breathing (WOB) | Pressure, Flow | Integrated calculation tracing the path coordinates of individual pressure-volume loops ($\int P \, dV$). |


## **Data Dictionary Provided:**

| Column Name | Data Type / Format | Description |
| :--- | :--- | :--- |
| `experiment_id` | String / UUID | Unique ID of the experiment |
| `patient_id` | String / Int | Unique patient ID |
| `timestamp` | Timestamp / Float | When it happened |
| `heart_rate` | Float / Int | Calculated HR from PPG waveforms |
| `lung_volume` | Float | Volume of the breath |
| `t_htu` | Float | Temperature of the air in the tube |
| `h_htu` | Float | Humidity of the air in the tube |
| `t_bmpA` | Float | Temperature of sensor Pa |
| `p_bmpA` | Float | Pressure at point A |
| `t_bmpB` | Float | Temperature of sensor Pb |
| `p_bmpB` | Float | Pressure at point B |
| `t_bmpC` | Float | Temperature of sensor Pc |
| `p_bmpC` | Float | Pressure at point C |
| `o2` | Float | Percent O2 being breathed out |
| `co2` | Float / Int | PPM CO2 being breathed out |
| `imu_x` | Float | Acceleration in X direction |
| `imu_y` | Float | Acceleration in Y direction |
| `imu_z` | Float | Acceleration in Z direction |
| `ppg_ir` | Float / Int | Raw waveform from the IR sensor of the PPG |
| `ppg_red` | Float / Int | Raw waveform from the RED sensor of the PPG |
| `bmi_gyro_x` | Float | Roll in x |
| `bmi_gyro_y` | Float | Roll in y |
| `bmi_gyro_z` | Float | Roll in z |
| `ppg_spo2` | Float | Blood oxygen saturation calculated from PPG |
| `hrv_sdnn` | Float | Heart rate variability from PPG |
| `bat_state` | String / Boolean | Was device plugged in or on battery |
| `source` | String / Boolean | Data came in cleanly |



# Import libraries

In [24]:
import mlflow
import pandas as pd
from mlflow.exceptions import MlflowException
import os
import numpy as np
import matplotlib as mpl

In [26]:
col_name = ['Feature_Name', 'Description', 'Theorical_Calculation', 'Type']
data_dict = pd.read_excel('../data/AeroCardia Data Dictionary Basic.xlsx', names=col_name)
data_dict.head(5)

,Feature_Name,Description,Theorical_Calculation,Type
0,Alveolar Ventilation (VA),Tidal Volume minus dead space and respiratory ...,V˙A=(VT−VD)×RR\dot V_A = (V_T - V_D)\times RR,Pulmonary
1,Baroreflex sensitivity (BRS),Coupling of HRV and beat-to-beat pressure esti...,ΔHR/ΔP\Delta HR / \Delta P slope,Cardiac
2,Breathing Efficiency,From Tidal Volume and respiratory rate you can...,Efficiency = VO₂ / Ventilation,Pulmonary
3,Breathing Frequency and Regularity,Timing and regularity of breaths. Irregular ca...,Frequency analysis of pressure waveform,Pulmonary
4,Cardiac Output (CO),HR and estimated SV can be used to calculate c...,CO=SV×HRCO = SV \times HR,Cardiac


#### Load Full Data-set

In [27]:
df_1 = pd.read_csv('../data/a595d7b2-4221-4961-a8b9-cf72eb07af14_samples.csv')
df_2 = pd.read_csv('../data/3ae687a7-fb94-4e75-a6ed-ddb3e00e132a_samples.csv')
df_3 = pd.read_csv('../data/743e9adb-3e6a-4f21-be67-76f8174fd899_samples.csv')
df_4 = pd.read_csv('../data/54032ac3-52b9-4af7-872f-2a6809af931c_samples.csv')
df_5 = pd.read_csv('../data/863b3091-8ee3-4e7e-bd32-36792ecf6835_samples.csv')
df_6 = pd.read_csv('../data/f4ea9b0c-5aba-475d-9dc4-01607c751b66_samples.csv')
df_7 = pd.read_csv('../data/ea9bdde1-f8cc-416c-a8d2-9df30757e0a9_samples.csv')
df_8 = pd.read_csv('../data/6b28f23a-b0a5-48f8-8058-6bd1ad9d7bff_samples.csv')
df_9 = pd.read_csv('../data/f4ea9b0c-5aba-475d-9dc4-01607c751b66_samples.csv')
df_10 = pd.read_csv('../data/5253bfbc-b99b-41af-b0b8-4ffa3a7c4362_samples.csv')
df_11 = pd.read_csv('../data/53056c36-b725-42a8-9d17-aefac2eb80f9_samples.csv')

df_final = pd.concat([df_1, df_2, df_3, df_4, df_5, df_6, df_7, df_8, df_9, df_10, df_11], ignore_index=True)
# Feature Engineering
# We are creating two more features  date and time
df_final['Date'] = df_final['timestamp'].map(lambda x : x.split('T')[0])
df_final['Time'] = df_final['timestamp'].map(lambda x : x.split('T')[1])

df_final.drop(columns=['experiment_id', 'patient_id', 'timestamp'], inplace=True)

df_final['Date'] = pd.to_datetime(df_final['Date'])
df_final.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 118428 entries, 0 to 118427
Data columns (total 26 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   heart_rate   118428 non-null  int64         
 1   lung_volume  0 non-null       float64       
 2   t_htu        118428 non-null  int64         
 3   h_htu        118428 non-null  int64         
 4   t_bmpA       118428 non-null  float64       
 5   p_bmpA       118428 non-null  float64       
 6   t_bmpB       118428 non-null  float64       
 7   p_bmpB       118428 non-null  float64       
 8   t_bmpC       118428 non-null  float64       
 9   p_bmpC       118428 non-null  float64       
 10  o2           118428 non-null  float64       
 11  co2          118428 non-null  int64         
 12  imu_x        118428 non-null  float64       
 13  imu_y        118428 non-null  float64       
 14  imu_z        118428 non-null  float64       
 15  ppg_ir       118428 non-null  int6

In [28]:
#df_final.Date.value_counts()
df_final.Date.value_counts().sort_values( ascending=False)

Date
2026-05-05    52624
2026-04-30    38509
2026-04-27    23818
2026-05-07     1615
2026-05-12     1096
2026-05-08      614
2026-05-13      152
Name: count, dtype: int64

### Missing values


In [5]:
missing_summary = df_final.isna().sum().sort_values(ascending=False)
missing_pct = missing_summary / len(df_final)
missing_pct

lung_volume    1.0
heart_rate     0.0
Date           0.0
source         0.0
bat_state      0.0
hrv_sdnn       0.0
ppg_spo2       0.0
bmi_gyro_z     0.0
bmi_gyro_y     0.0
bmi_gyro_x     0.0
ppg_red        0.0
ppg_ir         0.0
imu_z          0.0
imu_y          0.0
imu_x          0.0
co2            0.0
o2             0.0
p_bmpC         0.0
t_bmpC         0.0
p_bmpB         0.0
t_bmpB         0.0
p_bmpA         0.0
t_bmpA         0.0
h_htu          0.0
t_htu          0.0
Time           0.0
dtype: float64

### Duplicates

In [6]:
# Duplicate Values:
df_final.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 118428 entries, 0 to 118427
Data columns (total 26 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   heart_rate   118428 non-null  int64         
 1   lung_volume  0 non-null       float64       
 2   t_htu        118428 non-null  int64         
 3   h_htu        118428 non-null  int64         
 4   t_bmpA       118428 non-null  float64       
 5   p_bmpA       118428 non-null  float64       
 6   t_bmpB       118428 non-null  float64       
 7   p_bmpB       118428 non-null  float64       
 8   t_bmpC       118428 non-null  float64       
 9   p_bmpC       118428 non-null  float64       
 10  o2           118428 non-null  float64       
 11  co2          118428 non-null  int64         
 12  imu_x        118428 non-null  float64       
 13  imu_y        118428 non-null  float64       
 14  imu_z        118428 non-null  float64       
 15  ppg_ir       118428 non-null  int6

### Outliers

In [7]:
df_final.head(10)

,heart_rate,lung_volume,t_htu,h_htu,t_bmpA,p_bmpA,t_bmpB,p_bmpB,t_bmpC,p_bmpC,...,ppg_red,bmi_gyro_x,bmi_gyro_y,bmi_gyro_z,ppg_spo2,hrv_sdnn,bat_state,source,Date,Time
0,0,NaN,0,0,22.655457,101591.031250,22.760147,101583.765625,22.966232,101641.718750,...,0,8.605957,11.779785,-0.976562,0,0.0,0,full,2026-04-27,17:22:50.680Z
1,0,NaN,0,0,22.656982,101590.968750,22.759003,101583.718750,22.967834,101641.718750,...,0,12.939453,17.578125,15.747070,0,0.0,0,full,2026-04-27,17:22:50.755Z
2,0,NaN,0,0,22.658386,101590.953125,22.759048,101583.687500,22.971680,101641.656250,...,0,8.483887,24.963379,41.870117,0,0.0,0,full,2026-04-27,17:22:50.756Z
3,0,NaN,0,0,22.658386,101590.953125,22.759048,101583.687500,22.972198,101641.671875,...,0,8.483887,24.963379,41.870117,0,0.0,0,full,2026-04-27,17:22:50.767Z
4,0,NaN,0,0,22.658386,101590.953125,22.759048,101583.687500,22.974411,101641.671875,...,0,8.483887,24.963379,41.870117,0,0.0,0,full,2026-04-27,17:22:50.799Z
5,0,NaN,0,0,22.658386,101590.953125,22.759048,101583.687500,22.976440,101641.671875,...,0,8.483887,24.963379,41.870117,0,0.0,0,full,2026-04-27,17:22:50.832Z
6,0,NaN,0,0,22.658386,101590.953125,22.759048,101583.687500,22.977539,101641.703125,...,0,8.483887,24.963379,41.870117,0,0.0,0,full,2026-04-27,17:22:50.844Z
7,0,NaN,0,0,22.658386,101590.953125,22.759048,101583.687500,22.979279,101641.671875,...,0,8.483887,24.963379,41.870117,0,0.0,0,full,2026-04-27,17:22:50.873Z
8,0,NaN,0,0,22.658646,101590.843750,22.760529,101583.593750,22.980026,101641.640625,...,0,8.117676,3.479004,-12.756348,0,0.0,0,full,2026-04-27,17:22:50.889Z
9,0,NaN,0,0,22.658646,101590.843750,22.760773,101583.578125,22.981598,101641.609375,...,0,11.901855,5.310059,-23.498535,0,0.0,0,full,2026-04-27,17:22:50.917Z


In [8]:
num_cols = df_final.select_dtypes(include=['int64', 'float64']).columns
only_num = df_final[num_cols].describe().T.drop(columns=['count'], axis=1)
only_num



,mean,std,min,25%,50%,75%,max
heart_rate,62.402304,40.225123,0.000000,0.000000,81.000000,81.000000,169.000000
lung_volume,NaN,NaN,NaN,NaN,NaN,NaN,NaN
t_htu,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
h_htu,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
t_bmpA,25.445551,1.624846,0.000000,24.557465,24.859924,26.825348,28.245346
p_bmpA,99779.872106,3692.958920,0.000000,99164.484375,99831.125000,99876.562500,101632.390625
t_bmpB,25.734511,1.645225,0.000000,24.900391,25.129532,27.063812,28.634933
p_bmpB,99792.011717,3687.235936,0.000000,99226.656250,99781.484375,99829.687500,101595.828125
t_bmpC,34.306860,2.940765,22.966232,32.855640,35.734329,36.157242,39.197266
p_bmpC,99930.821299,935.940389,99136.312500,99163.562500,99869.109375,99919.781250,101650.515625


In [9]:
for i in only_num.itertuples():
    print(f"* {i.Index}, Mean: {i.mean}, Std: {i.std}, Min: {i.min}, 25% {i._4}, 50% {i._5} 75% {i._6} Max: {i.max}")


# Pandas(Index='heart_rate', mean=52.213230221252424, std=48.9168980730271, min=0.0, _4=0.0, _5=73.0, _6=95.0, max=169.0)


* heart_rate, Mean: 62.402303509305234, Std: 40.22512293768779, Min: 0.0, 25% 0.0, 50% 81.0 75% 81.0 Max: 169.0
* lung_volume, Mean: nan, Std: nan, Min: nan, 25% nan, 50% nan 75% nan Max: nan
* t_htu, Mean: 0.0, Std: 0.0, Min: 0.0, 25% 0.0, 50% 0.0 75% 0.0 Max: 0.0
* h_htu, Mean: 0.0, Std: 0.0, Min: 0.0, 25% 0.0, 50% 0.0 75% 0.0 Max: 0.0
* t_bmpA, Mean: 25.445551449096072, Std: 1.6248459669960282, Min: 0.0, 25% 24.557464599609375, 50% 24.85992431640625 75% 26.825347900390625 Max: 28.245346069335938
* p_bmpA, Mean: 99779.87210583646, Std: 3692.958920216041, Min: 0.0, 25% 99164.484375, 50% 99831.125 75% 99876.5625 Max: 101632.390625
* t_bmpB, Mean: 25.734511492901163, Std: 1.6452253262823282, Min: 0.0, 25% 24.900390625, 50% 25.129531860351562 75% 27.063812255859375 Max: 28.634933471679688
* p_bmpB, Mean: 99792.0117172987, Std: 3687.235935767738, Min: 0.0, 25% 99226.65625, 50% 99781.484375 75% 99829.6875 Max: 101595.828125
* t_bmpC, Mean: 34.30685964211926, Std: 2.940765089254483, Min: 22

# Observations:

* heart_rate, Mean: 52.213230221252424, Std: 48.9168980730271, Min: 0.0, 25% 0.0, 50% 73.0 75% 95.0 Max: 169.0
    * Need to **check the outlier that is 169**
* lung_volume, Mean: nan, Std: nan, Min: nan, 25% nan, 50% nan 75% nan Max: nan **NoData**
* t_htu, Mean: 0.0, Std: 0.0, Min: 0.0, 25% 0.0, 50% 0.0 75% 0.0 Max: 0.0 **NoData**
* h_htu, Mean: 0.0, Std: 0.0, Min: 0.0, 25% 0.0, 50% 0.0 75% 0.0 Max: 0.0 **NoData**
* t_bmpA, Mean: 26.22855357183371, Std: 1.44404408475087, Min: 10.763397216796877, 25% 25.236740112304688, 50% 26.782730102539062 75% 27.392013549804688 Max: 28.245346069335938
* p_bmpA, Mean: 100520.2190633674, Std: 845.4173776076566, Min: 99739.046875, 25% 99852.8125, 50% 99873.375 75% 101591.34375 Max: 101632.390625
* t_bmpB, Mean: 26.46238198270312, Std: 1.5238347755412394, Min: 7.757537841796875, 25% 25.504196166992188, 50% 26.904205322265625 75% 27.702804565429688 Max: 28.634933471679688
* p_bmpB, Mean: 100490.8790456985, Std: 860.9888996590682, Min: 99633.375, 25% 99815.484375, 50% 99826.953125 75% 101583.546875 Max: 101595.828125
* t_bmpC, Mean: 32.825962254487216, Std: 3.3365525358637047, Min: 22.966232299804688, 25% 31.01636505126953, 50% 33.154632568359375 75% 35.003173828125 Max: 38.09922790527344
* p_bmpC, Mean: 100563.29975271551, Std: 847.8582102014689, Min: 99830.21875, 25% 99899.0625, 50% 99916.75 75% 101639.90625 Max: 101650.515625
* o2, Mean: 19.672600252845946, Std: 7.752680153839962, Min: 7.745759963989258, 25% 17.965499877929688, 50% 19.36417007446289 75% 20.752689361572266 Max: 218.42434692382812
* co2, Mean: 0.0, Std: 0.0, Min: 0.0, 25% 0.0, 50% 0.0 75% 0.0 Max: 0.0
* imu_x, Mean: -0.24468705940543425, Std: 0.3347730296248606, Min: -1.835693359375, 25% -0.4287109375, 50% -0.23486328125 75% -0.0673828125 Max: 1.6875
* imu_y, Mean: 0.7208667889593394, Std: 0.4984657417079175, Min: -1.5634765625, 25% 0.731689453125, 50% 0.900146484375 75% 0.96044921875 Max: 2.863037109375
* imu_z, Mean: 0.026304412175401512, Std: 0.27055918454605477, Min: -1.166748046875, 25% -0.0849609375, 50% -0.0107421875 75% 0.080810546875 Max: 1.96435546875
* ppg_ir, Mean: 135790.76695493126, Std: 84069.68987800155, Min: 0.0, 25% 0.0, 50% 181953.0 75% 193713.5 Max: 240591.0
* ppg_red, Mean: 119436.07555313107, Std: 74664.34708998272, Min: 0.0, 25% 0.0, 50% 159781.0 75% 169138.0 Max: 223621.0
* bmi_gyro_x, Mean: -0.03298681531729828, Std: 34.25159419408864, Min: -697.1435546875, 25% -1.15966796875, 50% 0.18310546875 75% 1.64794921875 Max: 834.3505859375
* bmi_gyro_y, Mean: 0.3308709288560736, Std: 30.33746359647714, Min: -475.5859375, 25% -1.8310546875, 50% 0.244140625 75% 1.8310546875 Max: 471.8017578125
* bmi_gyro_z, Mean: -0.0014258216743947246, Std: 30.608400020459083, Min: -384.70458984375, 25% -1.953125, 50% 0.18310546875 75% 1.953125 Max: 453.43017578125
* ppg_spo2, Mean: 54.70648354645659, Std: 49.18190170462598, Min: 0.0, 25% 0.0, 50% 99.0 75% 99.0 Max: 99.0
* hrv_sdnn, Mean: 202.94208930319124, Std: 164.641139219529, Min: 0.0, 25% 0.0, 50% 264.8 75% 308.8 Max: 675.3
* bat_state, Mean: 0.0, Std: 0.0, Min: 0.0, 25% 0.0, 50% 0.0 75% 0.0 Max: 0.0

####
Questions:

- Ask Matthew what are the unites used in the metrics
- Data Dictionary of the dataset



# Define the KPI for "System Consistency"

We need to know what a "good" device test looks like.
- Does consistency mean low variance in the pressure sensors at rest?
- Does it mean the ppg_spo2 calculation doesn't drop signal when imu_x/y/z (movement) spikes?
- We need to define a target variable for "reliability."

# Map out the Physics and Biology Equations
We need to map exactly which raw columns feed into which planned features.

For example:

- Respiratory Rate (RR): Can we derive this using peak-detection on lung_volume or p_bmpA/B/C over time?
- Work of breathing: We will need to integrate the pressure-volume loop (using p_bmpA and lung_volume). This requires calculus (integration) over the time-series arrays.

## Respiratory Mechanics (Calculus & Peak Detection)

* These features rely on your volume (lung_volume) and pressure sensors (p_bmpA/B/C). we will be using numerical differentiation and integration here.

| Planned Feature | Raw Data Required | Mathematical Approach |
| :--- | :--- | :--- |
| **Respiratory Rate (RR)** | `lung_volume` or `p_bmpA` | Peak detection (e.g., `scipy.signal.find_peaks`) on the volume or pressure wave to count breaths per minute. |
| **Flow Rate** | `lung_volume`, `timestamp` | First derivative of volume over time. $$Flow = \frac{dV}{dt}$$ |
| **Work of Breathing** | `p_bmpA`, `lung_volume` | Integration of the pressure-volume loop for a single breath. $$WOB = \int P \, dV$$ |
| **Lung Compliance** | `lung_volume`, `p_bmpA` | Change in volume divided by change in pressure. $$C = \frac{\Delta V}{\Delta P}$$ |
| **Tidal Volume (VT)** | `lung_volume` | Maximum volume minus minimum volume for a single identified breath cycle. |


## Gas Exchange & Metabolism (Flow-Synchronized Integration)

- This group is tricky because gas sensors (o2, co2) often have a slight physical delay compared to the flow or volume sensors. we will need to align the time-series arrays perfectly before doing the math.

| Planned Feature | Raw Data Required | Mathematical Approach |
| :--- | :--- | :--- |
| **VCO2 (CO2 Output)** | `co2`, Flow Rate, `timestamp` | Integrate the product of exhaled flow and CO2 fraction over a breath. |
| **VO2 (O2 Uptake)** | `o2`, Flow Rate, `timestamp` | Integrate the difference between inhaled O2 and exhaled O2 multiplied by flow. |
| **RER** | VCO2, VO2 | Simple ratio: $$RER = \frac{\text{VCO2}}{\text{VO2}}$$ Tells us if the body is burning fats vs. carbs. |
| **Ventilatory Eq. (VE/VO2)** | `lung_volume`, `timestamp`, VO2 | Minute ventilation (total volume exhaled per minute) divided by VO2. |

## Cardiovascular & Kinematics (Signal Processing & Statistics)

- we already have heart_rate, ppg_spo2, and hrv_sdnn coming in cleanly, which saves us from having to write complex PPG waveform algorithms.

| Planned Feature | Raw Data Required | Mathematical Approach |
| :--- | :--- | :--- |
| **Physical Activity (PAL)** | `imu_x/y/z` | Calculate the rolling variance or signal magnitude area (SMA) of acceleration. |
| **Posture & Gait** | `bmi_gyro_x/y/z`, `imu_x/y/z` | Use tilt angles derived from gravity vectors on the IMU to classify posture (upright vs prone). |
| **Stroke Volume (SV) Est.** | `ppg_ir`, `ppg_red` | *Advanced*: Requires pulse contour analysis (measuring the area under the systolic portion of the raw PPG wave). |
| **Cardiac Output (CO)** | HR, SV | $$CO = HR \times SV$$ |



# Time-Series Alignment Strategy
Sensor data is messy. Your PPG might sample at 100Hz, your temperature sensor at 1Hz, and your O2 sensor at 10Hz. We need a strategy for resampling, interpolating, and aligning these timestamps into a clean Pandas DataFrame so we can run our models.